core-bundle/src/Resources/contao/dca/tl_article.php

```diff
@@ -773,17 +773,12 @@ public function cutArticle($row, $href, $label, $title, $icon, $attributes)
 	 */
 	public function pasteArticle(DataContainer $dc, $row, $table, $cr, $arrClipboard=null)
 	{
-		$imagePasteAfter = Image::getHtml('pasteafter.svg', sprintf($GLOBALS['TL_LANG'][$dc->table]['pasteafter'][1], $row['id']));
-		$imagePasteInto = Image::getHtml('pasteinto.svg', sprintf($GLOBALS['TL_LANG'][$dc->table]['pasteinto'][1], $row['id']));
+		@trigger_error('Deprecated since Contao 4.10, to be removed in Contao 5. Use Contao\CoreBundle\EventListener\DataContainer\ContentCompositionListener::renderArticlePasteButton instead.', E_USER_DEPRECATED);
 
-		if ($table == $GLOBALS['TL_DCA'][$dc->table]['config']['ptable'])
-		{
-			return ($row['type'] == 'root' || !$this->User->isAllowed(BackendUser::CAN_EDIT_ARTICLE_HIERARCHY, $row) || $cr) ? Image::getHtml('pasteinto_.svg') . ' ' : '<a href="' . $this->addToUrl('act=' . $arrClipboard['mode'] . '&amp;mode=2&amp;pid=' . $row['id'] . (!is_array($arrClipboard['id']) ? '&amp;id=' . $arrClipboard['id'] : '')) . '" title="' . StringUtil::specialchars(sprintf($GLOBALS['TL_LANG'][$dc->table]['pasteinto'][1], $row['id'])) . '" onclick="Backend.getScrollOffset()">' . $imagePasteInto . '</a> ';
-		}
-
-		$objPage = PageModel::findById($row['pid']);
-
-		return (($arrClipboard['mode'] == 'cut' && $arrClipboard['id'] == $row['id']) || ($arrClipboard['mode'] == 'cutAll' && in_array($row['id'], $arrClipboard['id'])) || !$this->User->isAllowed(BackendUser::CAN_EDIT_ARTICLE_HIERARCHY, $objPage->row()) || $cr) ? Image::getHtml('pasteafter_.svg') . ' ' : '<a href="' . $this->addToUrl('act=' . $arrClipboard['mode'] . '&amp;mode=1&amp;pid=' . $row['id'] . (!is_array($arrClipboard['id']) ? '&amp;id=' . $arrClipboard['id'] : '')) . '" title="' . StringUtil::specialchars(sprintf($GLOBALS['TL_LANG'][$dc->table]['pasteafter'][1], $row['id'])) . '" onclick="Backend.getScrollOffset()">' . $imagePasteAfter . '</a> ';
+		return System::getContainer()
+			->get(ContentCompositionListener::class)
+			->renderArticlePasteButton($dc, $row, $table, $cr, $arrClipboard)
+		;
```

Comment:
Why are we keeping this function? We have already removed whole `tl_*` classes from DCAs in the past. So I don't think it is necessary to keep that?

Reply:
As discussed on Slack, we always try to keep BC with methods that have been here a while, because noone knows who does _bad things_ with them 🤷